v1/plugin/plugin.go

```diff
@@ -328,5 +384,186 @@ func startPlugin(srv server, m *meta, p *pluginProxy) int {
 	go p.HeartbeatWatch()
 	// TODO(danielscottt): exit code
 	<-p.halt
-	return 0
+
+	return nil
+}
+
+// GetPluginType converts a pluginType to a string
+// describing what type of plugin this is
+func getPluginType(plType pluginType) string {
+	switch plType {
+	case collectorType:
+		return "collector"
+
+	case publisherType:
+		return "publisher"
+
+	case processorType:
+		return "processor"
+	}
+	return ""
+}
+
+// GetPluginType converts a pluginType to a string
+// as described in snap/control/plugin/plugin.go
+func getRPCType(i int) string {
+	if i == 0 {
+		return "NativeRPC (0)"
+	} else if i == 2 {
+		return "GRPC (2)"
+	}
+	return "Not a valid RPC Type"
+}
+
+func showDiagnostics(m meta, p *pluginProxy, c Config) error {
+	defer timeTrack(time.Now(), "showDiagnostics")
+	printRuntimeDetails(m)
+	err := printConfigPolicy(p, c)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	met, err := printMetricTypes(p, c)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	err = printCollectMetrics(p, met)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	printContactUs()
+	return nil
+
+}
+
+func printMetricTypes(p *pluginProxy, conf Config) ([]Metric, error) {
+	defer timeTrack(time.Now(), "printMetricTypes")
+	met, err := p.plugin.(Collector).GetMetricTypes(conf)
+	if err != nil {
+		return nil, fmt.Errorf("! Error in the call to GetMetricTypes: \n%v", err)
+	}
+	//apply any config passed in to met so that
+	//CollectMetrics can see the config for each metric
+	for i := range met {
+		met[i].Config = conf
+	}
+
+	fmt.Println("Metric catalog will be updated to include: ")
+	for _, j := range met {
+		fmt.Printf("    Namespace: %v \n", j.Namespace.String())
+	}
+	return met, nil
+}
+
+func printConfigPolicy(p *pluginProxy, conf Config) error {
+	defer timeTrack(time.Now(), "printConfigPolicy")
+	requiredConfigs := ""
+	cPolicy, err := p.plugin.(Collector).GetConfigPolicy()
+	if err != nil {
+		return err
+	}
+	fmt.Println("Config Policy:")
+	if len(cPolicy.stringRules) > 0 {
+		for k, v := range cPolicy.stringRules {
+			fmt.Printf("    Namespace: %-30v", k)
+			for key, val := range v.Rules {
```

Comment:
This could be broken into a method to be reused, but it would be more work than it is worth. Since v in the first loop and val in the second loop are different types for each block, it would be difficult to make a generic method that could handle that nicely.  This way, although it is repeating code, is more simple.